sunpy/data/data_manager/manager.py

```diff
@@ -0,0 +1,158 @@
+
+    Parameters
+    ----------
+    cache: `sunpy.data.data_manager.cache.Cache`
+        Cache object to be used by `~sunpy.data.data_manager.manager.DataManager`.
+    """
+
+    def __init__(self, cache):
+        self._cache = cache
+
+        self._file_cache = {}
+
+        self._skip_hash_check = False
+        self._skip_file = {}  # Dict[str, str]
+
+    def require(self, name, urls, sha_hash):
+        """
+        Decorator for informing the data manager about the requirement of
+        a file by a function.
+
+        Parameters
+        ----------
+        name: `str`
+            The name to reference the file with.
+        urls: `list` or `str`
+            A list of urls to download the file from.
+        sha_hash: `str`
+            SHA-1 hash of file.
+        """
+        if isinstance(urls, str):
+            urls = [urls]
+        def decorator(func):
+            @functools.wraps(func)
+            def wrapper(*args, **kwargs):
+                replace = self._skip_file.get(name, None)
+                err_message = "Hash does not match."
+                if replace:
+                    if replace['uri'].startswith('file://'):
+                        file_path = replace['uri'][len('file://'):]
+                        file_hash = hash_file(file_path)
+                    else:
+                        file_path, file_hash, _ = self._cache._download_and_hash([replace['uri']])
+                    if replace['hash'] and file_hash != replace['hash']:
+                        # if hash provided to replace function doesn't match the hash of the file
+                        # raise error
+                        raise KeyError(err_message + " Hash provided to replace_file does not match hash of the file.")
+                elif self._skip_hash_check:
+                    file_path = self._cache.download(urls, redownload=True)
+                else:
+                    details = self._cache.get_by_hash(sha_hash)
+                    if not details:
+                        # In case we are matching by hash and file does not exist
+                        # That might mean the wrong hash is supplied to decorator
+                        # We match by urls to make sure that is not the case
+                        if self._cache_has_file(urls):
+                            raise KeyError(err_message + " Hash provided does not match the hash in database.")
+                        file_path = self._cache.download(urls)
+                        if hash_file(file_path) != sha_hash:
+                            # the hash of the file downloaded does not match the hash of the file downloaded
+                            # this means the file has changed on the server.
+                            # the function should be updated to use the new hash. Raise an error to notify.
+                            raise RuntimeError("Remote file on the server has changed. Update hash of the function.")
```

Comment:
The fact that we error if the hash changes on the remote server is a conscious if opinionated design decision here.

We want to discourage files from being changed in place on remote servers, and we want to ensure that if I run some function in aia 0.1 then it always returns the same result until I run it with aiapy 0.2. There is a way of overriding this for the users with the context manager, which allows people to work around this. (This probably should be mentioned in the error)

If you know your file is likely to change a lot, and this is the point, then you probably want to just use the download with cache functionality here (like we [now do in map](https://github.com/sunpy/sunpy/pull/3124/files#diff-9306ac2bb55014e8e339ffdedaab50c5R228)).

Reply:
> We want to discourage files from being changed in place on remote servers, and we want to ensure that if I run some function in aia 0.1 then it always returns the same result until I run it with aiapy 0.2.

OK I can sort of understand that. But then is the expectation that each time a file changes on a remote server, we do a new release to update the hash? Doesn't this also mean that previous versions will "stop working" (i.e. raise an exception) because the hash is not consistent with that on the remote server?

Reply:
This API isn't well suited to a file you expect to be changed in place. It's more designed for a versioned file (even one published on zenodo etc).

The `cache.download` function allows you to do a download without knowing the hash before hand, and then not have to redownload it every time, which is more suited to that usecase I think.